File: session_shuffle.go
// Reimplement session shuffling algorithms in Go.  See the equivalent Python
// script for more comprehensive commenting of how each of the shuffling
// algorithms work.
package main

import (
	"bufio"
	"flag"
	"fmt"
	"math/rand"
	"os"
	"strconv"
	"strings"
	"sort"
)

func main() {
	seed := flag.Int64("s", 1000, "Seed for random shuffling.")
	flag.Parse()
	rand.Seed(*seed)
	pa := flag.Args()
	if len(pa) > 1 {
		fmt.Fprintf(os.Stderr, "Too many positional arguments!")
		os.Exit(1)
	} else if len(pa) < 1 {
		fmt.Fprintf(os.Stderr, "Too few positional arguments!")
		os.Exit(1)
	}
	alg := strings.ToLower(pa[0])

	// read sessions from stdin
	sessions := []Session{}
	scanner := bufio.NewScanner(os.Stdin)
	for scanner.Scan() {
		fields := strings.Split(strings.TrimSpace(scanner.Text()), ",")
		node := fields[0]
		start, _ := strconv.Atoi(fields[1])
		end, _ := strconv.Atoi(fields[2])
		location := fields[3]
		sessions = append(sessions, Session{node, start, end, location})
	}

	switch alg {
	case "tn":
		sessions = tn(sessions)
	case "ln":
		sessions = ln(sessions)
	case "tl":
		sessions = tl(sessions)
	case "tlln":
		sessions = tlln(sessions)
	case "lntn":
		sessions = lntn(sessions)
	case "_":
		sessions = destroy_all(sessions)
	default:
		fmt.Fprintf(os.Stderr, "Unrecognized shuffle algorithm: %s",
			alg)
		os.Exit(1)
	}

	// Output the shuffled sessions.
	for _, s := range sessions {
		fmt.Println(strings.Join([]string{
			s.node,
			strconv.Itoa(s.start),
			strconv.Itoa(s.end),
			s.location},
			","))
	}
}

func tn(sessions []Session) []Session {
	for i, _ := range sessions {
		j := rand.Intn(i + 1)
		sessions[i].location, sessions[j].location =
			sessions[j].location, sessions[i].location
	}
	return sessions
}

func ln(sessions []Session) []Session {
	for i, _ := range sessions {
		j := rand.Intn(i + 1)
		sessions[i].start, sessions[j].start =
			sessions[j].start, sessions[i].start
		sessions[i].end, sessions[j].end =
			sessions[j].end, sessions[i].end
	}
	return sessions
}

func tl(sessions []Session) []Session {
	for i, _ := range sessions {
		j := rand.Intn(i + 1)
		sessions[i].node, sessions[j].node =
			sessions[j].node, sessions[i].node
	}
	return sessions
}

func tlln(sessions []Session) []Session {
	// group durations by location.
	loc_durs := make(map[string][]Duration)
	for _, s := range sessions {
		_, ok := loc_durs[s.location]
		if !ok {
			loc_durs[s.location] = []Duration{}
		}
		loc_durs[s.location] = append(loc_durs[s.location],
			Duration{s.start, s.end})
	}
	// in each group shuffle durations.
	// sort keys first to ensure determinism.
	var keys []string
	for k := range loc_durs {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	for _, k := range keys {
		durs := loc_durs[k]
		for i, _ := range durs {
			j := rand.Intn(i + 1)
			durs[i], durs[j] = durs[j], durs[i]
		}
	}
	// apply shuffled durations to the original sessions.
	for i, _ := range sessions {
		loc := sessions[i].location
		// two-step pop
		new_dur := loc_durs[loc][len(loc_durs[loc])-1]
		loc_durs[loc] = loc_durs[loc][:len(loc_durs[loc])-1]

		sessions[i].start, sessions[i].end = new_dur.start, new_dur.end
	}
	return sessions
}

func lntn(sessions []Session) []Session {
	// group locations by node.
	node_locs := make(map[string][]string)
	for _, s := range sessions {
		_, ok := node_locs[s.node]
		if !ok {
			node_locs[s.node] = []string{}
		}
		node_locs[s.node] = append(node_locs[s.node], s.location)
	}
	// in each group shuffle locations.
	// sort keys first to ensure determinism.
	var keys []string
	for k := range node_locs {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	for _, k := range keys {
		locs := node_locs[k]
		for i, _ := range locs {
			j := rand.Intn(i + 1)
			locs[i], locs[j] = locs[j], locs[i]
		}
	}
	// for _, locs := range node_locs {
	// 	for i, _ := range locs {
	// 		j := rand.Intn(i + 1)
	// 		locs[i], locs[j] = locs[j], locs[i]
	// 	}
	// }
	// apply shuffled locations to the original sessions.
	for i, _ := range sessions {
		node := sessions[i].node
		// two-step pop
		new_loc := node_locs[node][len(node_locs[node])-1]
		node_locs[node] = node_locs[node][:len(node_locs[node])-1]

		sessions[i].location = new_loc
	}
	return sessions
}

func destroy_all(sessions []Session) []Session {
	for i, _ := range sessions {
		j := rand.Intn(i + 1)
		sessions[i].node, sessions[j].node =
			sessions[j].node, sessions[i].node
	}
	for i, _ := range sessions {
		j := rand.Intn(i + 1)
		sessions[i].location, sessions[j].location =
			sessions[j].location, sessions[i].location
	}
	return sessions
}

type Session struct {
	node       string
	start, end int
	location   string
}

type Duration struct {
	start, end int
}

File: session_shuffle.py
""" Session shuffling algorithms.

Each session shuffling algorithm is implemented as a function with the same
signature.  That is, each algorithm takes as input a set of session records
which is an iterable where each element is a 4-element iterable of the form
<node, start, end, location> describing a session.

Each session shuffling algorithm also returns the same type of data
structure.  That data structure takes exactly the same form as the input
data -- an iterable where each element is a 4-element iterable of the
form <node, start, end, location> that is the product of performing the
relevant shuffling algorithm over the input data.


If called as main script:

stdin

    Comma-separated four-tuples, one tuple per line, of the form <node,
    start, end, location> describing a session.

stdout

    Comma-separated four-tuples, one tuple per line, of the input data
    after performing the selected shuffling algorithm (specified via
    positional parameter)

positional parameters

    A shuffling algorithm acronym.  Call script with -h for available options.

"""

import sys
import random
import argparse
from collections import defaultdict

def tn(sessions):
    """ Shuffle location.

    Correlations retained:

        * Time and Node (TN)

    Correlations destroyed:

        * Time and Location (TL)
        * Location and Node (LN)
    """
    aps = []
    AP_IDX = 3
    for s in sessions:
        aps.append(s[AP_IDX])
    random.shuffle(aps)
    out = []
    for s in sessions:
        out.append([s[0], s[1], s[2], aps.pop()])
    return out


def ln(sessions):
    """ Shuffle times.

    Correlations retained:

        * Location and Node (LN)

    Correlations destroyed:

        * Time and Location (TL)
        * Time and Node (TN)
    """
    time_pairs = []
    START_IDX, END_IDX = 1, 2
    for s in sessions:
        time_pairs.append([s[START_IDX], s[END_IDX]])
    random.shuffle(time_pairs)
    out = []
    for s in sessions:
        new_start, new_end = time_pairs.pop()
        out.append([s[0], new_start, new_end, s[3]])
    return out


def tl(sessions):
    """ Shuffle nodes.

    Correlations retained:

        * Time and Location (TL)

    Correlations destroyed:

        * Location and Node (LN)
        * Time and Node (TN)

    """
    macs = []
    MAC_IDX = 0
    for s in sessions:
        macs.append(s[MAC_IDX])
    random.shuffle(macs)
    out = []
    for s in sessions:
        mac = macs.pop()
        out.append([mac, s[1], s[2], s[3]])
    return out


def tlln(sessions):
    """ Group by location, shuffle time.

    Correlations retained:

        * Time and Location (TL)
        * Location and Node (LN)

    Correlations destroyed:

        * Time and Node (TN)

    """
    # key = ap, value = list of [start, end] two-pairs
    ap_times = defaultdict(list)
    AP_IDX, START_IDX, END_IDX = 3, 1, 2
    for s in sessions:
        ap, start, end = s[AP_IDX], s[START_IDX], s[END_IDX]
        ap_times[ap].append([start, end])
    for k in ap_times.keys():
        times = ap_times[k]
        random.shuffle(times)
        ap_times[k] = times
    out = []
    for s in sessions:
        ap = s[AP_IDX]
        new_start, new_end = ap_times[ap].pop()
        out.append([s[0], new_start, new_end, s[3]])
    return out


def lntn(sessions):
    """ Group by node, shuffle location.

    Correlations retained:

        * Location and Node (LN)
        * Time and Node (TN)

    Correlations destroyed:

        * Time and Location (TL)
    
    """
    mac_aps = defaultdict(list)
    MAC_IDX, AP_IDX = 0,3
    for s in sessions:
        mac, ap = s[MAC_IDX], s[AP_IDX]
        mac_aps[mac].append(ap)
    for k in mac_aps.keys():
        aps = mac_aps[k]
        random.shuffle(aps)
        mac_aps[k] = aps
    out = []
    for s in sessions:
        mac = s[MAC_IDX]
        out.append([s[0], s[1], s[2], mac_aps[mac].pop()])
    return out


def tltn(sessions):
    """ Group by time, shuffle node.

    Correlations retained:

        * Time and Location (TL)
        * Time and Node (TN)

    Correlations destroyed:

        * Location and Node (LN)

    """
    # this will need some figuring out because time will need to be "bucketed"
    # somehow
    raise NotImplementedError('TLTD not implemented yet!')

def destroy_all(sessions):
    """Shuffle by node, shuffle by location.

    Correlations retained:

        * None

    Correlations destroyed:

        * Time and Location (TL)
        * Time and Node (TN)
        * Location and Node (LN)

    """
    nodes, locations = [], []
    NODE_IDX, LOC_IDX = 0, 3
    for s in sessions:
        nodes.append(s[NODE_IDX])
        locations.append(s[LOC_IDX])
    random.shuffle(nodes)
    random.shuffle(locations)
    out = []
    for s in sessions:
        node = nodes.pop()
        location = locations.pop()
        out.append([node, s[1], s[2], location])
    return out


if __name__ == "__main__":
    # configure arguments
    parser = argparse.ArgumentParser(
        description='Also see module-level documentation.')
    parser.add_argument('algorithm', type=str,
                        help="One of the null model abbreviations.")
    parser.add_argument('-s', '--random-seed', type=int, default=1000,
                        help='A random seed for shuffling')
    args = parser.parse_args()
    random.seed(args.random_seed)

    # read sessions from stdin
    sessions = []
    for line in sys.stdin:
        f = line.strip().split(',')
        sessions.append([f[0], int(f[1]), int(f[2]), f[3]])

    # shuffle (or don't if original)
    algorithm = args.algorithm.lower()
    if algorithm == 'original':
        for s in sessions:
            print ','.join(map(str, s))
    else:
        if algorithm == '_':
            algorithm = 'destroy_all'
        thismodule = sys.modules[__name__]
        alg_func = getattr(thismodule, algorithm)
        shuffled = alg_func(sessions)
        for s in shuffled:
            print ','.join(map(str, s))
